std/variant.d

```diff
@@ -253,7 +253,13 @@ private:
                 {
                     return 0;
                 }
-                static if (is(typeof(*zis < *rhsPA)))
+                static if (is(A == struct) && A.init.tupleof.length == 0)
+                {
+                    // The check above will always succeed if A is an empty struct.
```

Comment:
Why not `return 0;` here?

Reply:
Because it would be unreachable code - diagnosed with `-preview=fieldwise`.

Reply:
What about moving

```D
if (*rhsPA == *zis) { return 0; }
```

as two separate code duplications into the `else static if` and `else` branches? I guess we can afford some code duplication here.

That should remove the warning for the

```D
static if (is(A == struct) && A.init.tupleof.length == 0)
```

case right?

Reply:
This compiles for me without the flag:

```d
        static ptrdiff_t compare(A* rhsPA, A* zis, OpID selector)
        {
            static if (is(typeof(*rhsPA == *zis)))
            {
                static if (!(is(A == struct) && A.init.tupleof.length == 0))
                {
                    if (*rhsPA == *zis)
                        return 0;
                }
                static if (is(A == struct) && A.init.tupleof.length == 0)
                {
                    return 0;
                }
                else static if (is(typeof(*zis < *rhsPA)))
                {
                    // Many types (such as any using the default Object opCmp)
                    // will throw on an invalid opCmp, so do it only
                    // if the caller requests it.
                    if (selector == OpID.compare)
                        return *zis < *rhsPA ? -1 : 1;
                    else
                        return ptrdiff_t.min;
                }
                else
                {
                    // Not equal, and type does not support ordering
                    // comparisons.
                    return ptrdiff_t.min;
                }
            }
            else
            {
                // Type does not support comparisons at all.
                return ptrdiff_t.min;
            }
        }
```

Reply:
That would also need to check for user-defined `opEquals` inside of an empty struct.

Reply:
Can you elaborate on why you need to check for `opEquals`?

Reply:
The user-defined `opEquals` may have different semantics than the compiler generated one (e.g. `return false`
). Your proposal skips it and simply returns "equal".

Additionally, the warning disappears when the empty struct defines an `opEquals` (see https://run.dlang.io/is/rvo4Yt), which suggests that it breaks the early const-folding. That would trigger the `missing return`,... again

Reply:
> The user-defined `opEquals` may have different semantics than the compiler generated one (e.g. `return false`
> ). Your proposal skips it and simply returns "equal".

What about statically checking for presence of user-defined `opEquals` and avoid special handling then?

Have you asked @WalterBright how to proceed with these problems?

Reply:
> What about statically checking for presence of user-defined `opEquals` and avoid special handling then?

That could work
 
> Have you asked @WalterBright how to proceed with these problems?

No and I don't see a reason to do so. He rarely contributes to phobos and there's more imporant work only he can do (e.g. review backend pull requests)

Reply:
Ok, thanks.

Reply:
Yup, the bright side of the lack of leadership is that we have more decision freedom here than at DMD ;-)